### PR TITLE
chore(ZNTA-1414) add Jenkinsfile for pipeline builds

### PR DESCRIPTION
**Pull request by seanf**
TODO:
- [x] Ensure that chrome/chromedriver processes don't hang around as orphans/zombies if the build is aborted
- [x] Ensure that broken unit tests or checkstyle rules in, say, client are reported quickly (no waiting for ~complete build, or~ functional tests)
## Reviewer tasks

The following is based on the
[Zanata Development Guidelines](https://github.com/zanata/zanata-server/wiki/Development-Guidelines).
As a reviewer, you should check the guidelines regularly for updates, and just
to refresh your memory.

The reviewer can use this list for reference to make sure
they have considered all the important points.

Use the checkboxes or not, whatever works best for you.
- Code quality
  - [ ] Code passes style check (checkstyle/eslint)
  - [ ] Code is clear and concise
  - [ ] UI code is internationalised
  - [ ] Code has adequate comments where appropriate
  - Code is in an appropriate place
    - [ ] Classes are in appropriate packages
    - [ ] Code fits with class's responsibilities (SRP)
  - [ ] Configuration files are documented enough
  - If code is removed
    - [ ] No remaining code references the removed code
      - [ ] No orphaned rewrite rules
    - [ ] No orphaned config settings
    - [ ] No remaining comments refer to the removed code
    - [ ] Unused imports and libraries are removed
- Testing
  - [ ] New code has tests
  - [ ] Changed code has tests
  - [ ] All tests pass
  - [ ] Test coverage stable or increased
- Documentation
  - [ ] New features are documented
  - [ ] Docs removed for deleted features
  - [ ] Docs updated for all changed features
  - [ ] Developer/sysadmin docs updated if appropriate
- If there are new dependencies
  - [ ] Does each new dependency pass all the
     [Considerations for new dependencies](https://github.com/zanata/zanata-server/wiki/Development-Guidelines#new-technologies-and-dependencies)?

---

_This template can be updated in .github/PULL_REQUEST_TEMPLATE.md_

---

This change is [Reviewable](https://reviewable.io/reviews/zanata/zanata-platform/2)

ghprb plugin: [skip ci]

### CODE REVIEW

**Comment by seanf** [comment]
@definite I've made a start on porting the `addJenkinsfile` branch to work with zanata-platform here in this branch. Please ignore the old branch.

**Comment by seanf** [comment]
@definite 

I suggest we remove the option `-Dmaven.test.failure.ignore` to enable fail-fast (eg broken client test), but we will need to ensure that we still collect JUnit results before aborting the build.

I think we could either use `sh`'s `returnStatus` option, or just catch the exception thrown by `sh "mvnw ..."`:
  https://jenkins.io/doc/pipeline/steps/workflow-durable-task-step/#code-sh-code-shell-script

Also, can you remember the plan for the zombie chrome/driver processes, given that `run-clean.sh` can't be bulletproof? I think we were going to pkill chrome/chromedriver processes in case they were left over by a previous build, but how were we going to avoid killing processes which belong to another executor?

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+void setJUnitPrefix(prefix, files) {
+  // add prefix to qualified classname
+  sh "find . -path \"*/${file}\" -exec sed -i \"s/\\(<testcase .*classname=['\\\"]\\)\\([a-z]\\)/\\1${prefix}.\\2/g\" '{}' +"
+}
+
+void archiveTestFilesIfUnstable() {
+  // if tests have failed currentBuild.result will be 'UNSTABLE'
+  if (currentBuild.result != null) {
+    archive(
+        includes: '*/target/**/*.log,*/target/screenshots/**,**/target/site/jacoco/**,**/target/site/xref/**',
```
Comment (seanf) [inline Jenkinsfile]: @definite On reflection, I don't think we should archive jacoco results for failed tests.

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+      }
+    }
+  }catch(e){
+    currentBuild.result = 'UNSTABLE'
+    archiveTestFilesIfUnstable()
+    throw e
+  }finally{
+    notify.testResults(appserver.toUpperCase())
+    junit allowEmptyResults: true,
+        keepLongStdio: true,
```
Comment (seanf) [inline Jenkinsfile]: http://javadoc.jenkins-ci.org/junit/hudson/tasks/junit/JUnitParser.html#JUnitParser-boolean-boolean-

Having looked into it, I don't think we need the option keepLongStdio, for failsafe, or for surefire above. Could you remove them please @definite ?

File: Jenkinsfile (hunk 1 above)
Comment (seanf) [inline Jenkinsfile]: In fact, we shouldn't bother archiving them at all. The only thing we should do with them is generate coverage reports.

**Comment by definite** [comment]
@zanata-jenkins retest this please

**Comment by seanf** [comment]
Squashed to https://github.com/zanata/zanata-platform/pull/254

**Comment by seanf** [comment]
Test comment for ultrahook.

**Pull request closed without merging**